From d1f095167d4ad3445505fd01125dcda68d0908e2 Mon Sep 17 00:00:00 2001
From: Yin Tirui <yintirui@huawei.com>
Date: Thu, 13 Nov 2025 15:55:51 +0800
Subject: [PATCH 890/960] downstream: pgtable: add pte_clrhuge() implementation
 for arm64

Add pte_clrhuge() helper function for arm64 to convert a huge PTE
entry to regular PTE.

This function provides the inverse operation of pte_mkhuge() and is
needed when splitting huge page mappings, where individual PTE entries
are created from the original huge page mapping.

Signed-off-by: Yin Tirui <yintirui@huawei.com>
Conflicts:
	include/linux/pgtable.h
[Fanhua Li: context conflict]
Signed-off-by: Fanhua Li <lifanhua5@huawei.com>
Signed-off-by: Liu Shixin <liushixin2@huawei.com>
---
 arch/arm64/include/asm/pgtable.h | 9 +++++++++
 arch/x86/include/asm/pgtable.h   | 1 +
 include/linux/pgtable.h          | 5 +++++
 3 files changed, 15 insertions(+)

diff --git a/arch/arm64/include/asm/pgtable.h b/arch/arm64/include/asm/pgtable.h
index a2071ad54280..df8a1ddeec21 100644
--- a/arch/arm64/include/asm/pgtable.h
+++ b/arch/arm64/include/asm/pgtable.h
@@ -252,6 +252,15 @@ static inline pte_t pte_mkpresent(pte_t pte)
 	return set_pte_bit(pte, __pgprot(PTE_VALID));
 }
 
+static inline pte_t pte_clrhuge(pte_t pte)
+{
+	pteval_t mask = PTE_TYPE_MASK & ~PTE_VALID;
+	pteval_t val = PTE_TYPE_PAGE & ~PTE_VALID;
+
+	return __pte((pte_val(pte) & ~mask) | val);
+}
+#define pte_clrhuge pte_clrhuge
+
 static inline pmd_t pmd_mkcont(pmd_t pmd)
 {
 	return __pmd(pmd_val(pmd) | PMD_SECT_CONT);
diff --git a/arch/x86/include/asm/pgtable.h b/arch/x86/include/asm/pgtable.h
index 4e3fb2545c30..9131e774c536 100644
--- a/arch/x86/include/asm/pgtable.h
+++ b/arch/x86/include/asm/pgtable.h
@@ -510,6 +510,7 @@ static inline pte_t pte_clrhuge(pte_t pte)
 {
 	return pte_clear_flags(pte, _PAGE_PSE);
 }
+#define pte_clrhuge pte_clrhuge
 
 static inline pte_t pte_mkglobal(pte_t pte)
 {
diff --git a/include/linux/pgtable.h b/include/linux/pgtable.h
index cb210802b4e4..4b1e00ba7e5a 100644
--- a/include/linux/pgtable.h
+++ b/include/linux/pgtable.h
@@ -575,6 +575,11 @@ static inline pte_t ptep_get_and_clear_full(struct mm_struct *mm,
 }
 #endif
 
+#ifndef CONFIG_ARCH_SUPPORTS_HUGE_PFNMAP
+#ifndef pte_clrhuge
+#define pte_clrhuge(pte) (pte)
+#endif
+#endif
 
 /*
  * If two threads concurrently fault at the same page, the thread that
-- 
2.43.0

